### PR TITLE
[Feature] llama shardformer fp8 support

### LINKED ISSUE #5916: [Feature] support shardformer
Replace the dtype to fp8 in communication is useful For bandwidth restricted scenario. Please support the fp8 communication in Shardformer(SP and TP)

### PR DESCRIPTION
**Pull request by GuangyaoZhang**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs
- [x] I have installed pre-commit: `pip install pre-commit && pre-commit install`

## 🚨 Issue number

resolve #5916

## 📝 What does this PR do?

Add FP8 support in Llama shardformer and reach loss parity during pretraining task

截屏2024-07-31 13 50 52

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: tests/test_fp8/test_fp8_all_to_all_single.py [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+from torch.distributed.distributed_c10d import _get_default_group
+from torch.testing import assert_close
+
+from colossalai import launch
+from colossalai.accelerator import get_accelerator
+from colossalai.shardformer.layer._operation import _all_to_all_single
+from colossalai.testing import parameterize, rerun_if_address_is_in_use, spawn
+
+
+@parameterize("shape", [(1, 8, 16)])
```
Comment (flybird11111) [inline tests/test_fp8/test_fp8_all_to_all_single.py:12]: In this test, the first dimension of the input shape is preferably greater than 1.

File: colossalai/quantization/fp8.py [hunk 2]
```diff
@@ -49,20 +52,23 @@ def cast_from_fp8(inp: torch.Tensor, scale_inv: torch.Tensor, ret_type: torch.dt
...
-    if inp.dim() == 2:
+    if per_channel_scale:
         ret = scale_inv[:, None] * inp.float()
     else:
         ret = scale_inv * inp.float()
     return ret.to(ret_type)
 
 
-def all_reduce_fp8(tensor: torch.Tensor, fp8_format="e5m2", group=None) -> None:
+def all_reduce_fp8(tensor: torch.Tensor, fp8_format="e4m3", op="sum", group=None) -> None:
```
Comment (ver217) [inline colossalai/quantization/fp8.py:62]: Can we use `torch.distributed.distributed_c10d.ReduceOp` instead of string "sum"?

File: colossalai/shardformer/layer/_operation.py [hunk 3]
```diff
@@ -954,14 +1002,33 @@ def _reduce_scatter(input_, dim=1, process_group=None):
...
+        output_list = [torch.empty_like(input_list[0]) for _ in range(world_size)]
+        scale_list = [torch.ones(1, dtype=scale.dtype, device=input_.device) for _ in range(world_size)]
+        dist.all_to_all(output_list, input_list, group=group)
+        dist.all_gather(scale_list, scale, group=group)
+        cast_tensor_list = []
+        for output, scale in zip(output_list, scale_list):
+            output = output.view(fp8_type)
+            output = cast_from_fp8(output, scale, input_type)
+            cast_tensor_list.append(output)
+        output_list = cast_tensor_list
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:1021]: add a new function named all_to_all_fp8

File: colossalai/shardformer/layer/_operation.py [hunk 4]
```diff
@@ -973,8 +1040,24 @@ def _all_to_all_single(input_, seq_world_size, group, scatter_dim, gather_dim):
...
+        output = torch.empty_like(input_t)
+        scale_list = [torch.ones(1, dtype=scale.dtype, device=input_.device) for _ in range(seq_world_size)]
+        dist.all_to_all_single(output, input_t, group=group)
+        dist.all_gather(scale_list, scale, group=group)
+        cast_tensor_list = []
+        for output_part, scale in zip(output, scale_list):
+            output_part = output_part.view(fp8_type)
+            output_part = cast_from_fp8(output_part, scale, input_type)
+            cast_tensor_list.append(output_part)
+        output = torch.stack(cast_tensor_list, dim=0)
```
Comment (ver217) [inline colossalai/shardformer/layer/_operation.py:1057]: add a new function named all_to_all_single_fp8

File: tests/test_fp8/test_fp8_all_to_all_single.py (hunk 1 above)
Comment (GuangyaoZhang) [reply tests/test_fp8/test_fp8_all_to_all_single.py]: Fixed

File: colossalai/quantization/fp8.py (hunk 2 above)
Comment (GuangyaoZhang) [reply colossalai/quantization/fp8.py]: Fixed

File: colossalai/shardformer/layer/_operation.py (hunk 3 above)
Comment (GuangyaoZhang) [reply colossalai/shardformer/layer/_operation.py]: Fixed

File: colossalai/shardformer/layer/_operation.py (hunk 4 above)
Comment (GuangyaoZhang) [reply colossalai/shardformer/layer/_operation.py]: Fixed

**Review by ver217 — APPROVED** [review]
